tests/core/go_binary/pgo_test.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pgo_test
+
+import (
+	_ "embed"
+	"os"
+	"path"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel_testing"
+)
+
+//go:embed pgo.pprof
+var pgoProfile []byte
+
+func TestMain(m *testing.M) {
+	bazel_testing.TestMain(m, bazel_testing.Args{
+		Main: `
+-- src/BUILD.bazel --
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_test")
+
+go_binary(
+    name = "pgo_with_profile",
+    srcs = ["pgo.go"],
+    pgoprofile = ":pgo.pprof",
+)
+
+go_binary(
+    name = "pgo_without_profile",
+    srcs = ["pgo.go"],
+)
+
+go_test(
+    name = "compare_results_test",
+    data = [
+        ":pgo_with_profile",
+        ":pgo_without_profile",
+    ],
+    srcs = ["compare_results_test.go"],
+    deps = ["@io_bazel_rules_go//go/runfiles:go_default_library"],
+)
+
+-- src/pgo.go --
+package main
+
+import "fmt"
+
+func main() {
+  fmt.Println("Did you know that profile guided optimization was added to the go compiler in go version 1.20?")
+}
+
+-- src/compare_results_test.go --
+package compare_results_test
+
+import (
+    "crypto/sha256"
+    "testing"
+
+    "github.com/bazelbuild/rules_go/go/runfiles"
+)
+
+func TestResultsAreSame(t *testing.T) {
+    r, err := runfiles.New()
+    if err != nil {
+        t.Fatal(err)
+    }
+
+    f1, err := r.ReadFile("__main__/src/pgo_with_profile_/pgo_with_profile")
+    if err != nil {
+        t.Fatal(err)
+    }
+    f2, err := r.ReadFile("__main__/src/pgo_without_profile_/pgo_without_profile")
+    if err != nil {
+        t.Fatal(err)
+    }
+
+    a, b := sha256.Sum256(f1), sha256.Sum256(f2)
+
+    if a == b {
+        t.Fatal("outputs are equal when they should be different")
+    }
+}
+`,
+	})
+}
+
+func TestGoBinaryOutputWithPgoProfileDiffersFromGoBinaryWithoutPgoProfile(t *testing.T) {
```

Comment:
This test fails on windows, possibly due to the hardcoded runfile strings on lines 81 and lines 85. Is there a better way to read those files?

Reply:
You can pass in the correct paths via `go_test` `env` using `$(rlocationpath ...)` instead, which will also give you the correct path on Windows.

Reply:
Nevermind, it's because windows uses `.exe` file extensions for executables. :rofl: